Maximize button for resizable instruments

michaelgregorius opened the pull request:
Show the maximize button for resizable instruments.

Most other changes have the character of refactorings and code reorganizations.

Remove the negation in the if condition for resizable instruments to make the code better readable.

Only manipulate the system menu if the instrument is not resizable.

Add a TODO to the special code that sets a size.

Unfortunately there are still some artifacts for some reason, i.e. the title bar still seems to be shown for a maximized instrument window:

![7512-ArtifactsForMaximizedInstrumentWindow](https://github.com/user-attachments/assets/30ed0d3b-475b-4e31-a3a6-96c426d88fdc)

michaelgregorius commented:
This seems to be caused by LMMS' own `SubWindow` class because it works as expected if I directly add a QMdiSubWindow via the MDI area:
```
auto wid = new QWidget();
wid->setMinimumSize(QSize(200, 100));
wid->setWindowTitle("Window title");
m_workspace->addSubWindow(wid);
```

michaelgregorius commented:
Maximizing also works as expected when the instrument window is added as a "normal" `QMdiSubWindow`, i.e. not as a `SubWindow`:

![7512-InstrumentsWithQMdiSubWindows](https://github.com/user-attachments/assets/9396a8d8-3966-4a98-97ec-fb8b3744d060)

michaelgregorius commented:
This PR is now ready for review as I have managed to fix the problems with the `SubWindow` class so that it now looks like in the screenshot above:

![7512-InstrumentsWithQMdiSubWindows](https://github.com/user-attachments/assets/9396a8d8-3966-4a98-97ec-fb8b3744d060)

I have also added some other small improvements to the `SubWindow` class.

sakertooth commented:
@michaelgregorius, I have plans to revert a good chunk of #7453, which would affect this PR. I don't think we should keep moving forward with making SlicerT bigger/resizable for now until we properly rework the instrument window to support larger sizes the right way. Effect racks and (at the very least) Lv2 layouts do not render properly and the envelope tab is massively distorted.

We will always have another chance at making the instrument windows bigger, it's not like I am throwing that idea out completely. I just don't think its worth progressing further with this knowing the massive regressions we still have to deal with (and I do not like how resizability/bigger sizes for instruments was implemented).

michaelgregorius commented:
@sakertooth, this pull request can be reviewed and merged nevertheless. The changes in `src/gui/instrument/InstrumentTrackWindow.cpp` prepare the instrument window to do "the right thing" once it is resizable and the changes in `SubWindow` fix a problem that exists in the class independent of instrument resizing.

So, I'd appreciate a review.

JohannesLorenz commented on src/gui/instrument/InstrumentTrackWindow.cpp:
```diff
@@ -285,24 +285,30 @@ InstrumentTrackWindow::InstrumentTrackWindow( InstrumentTrackView * _itv ) :
+
+	if (m_instrumentView->isResizable())
+	{
+		// TODO As of writing SlicerT is the only resizable instrument. Is this code specific to SlicerT?
```
Note, with #7201 , there will be a lot more resizable instruments

JohannesLorenz replied:
I assume that the diff between `InstrumentView` (which contains SlicerT OR ZynAddSubFx OR TripleOsc etc) and the whole subwindow is just not dependent on the instrument (SlicerT), but it is rather:

* The additional width is the small border left and right of SlicerT to form the whole SubWindow
* The larger 208 height is for the PianoRoll below and the knobs etc above?

It might be interesting (but also terrifying) to check if the calculation matches the numbers from PianoView, InstrumentTrackWindow etc. :laughing:

JohannesLorenz commented:
As my PR from #7524 does not fully work and conflict with your PR, I will try to use your PR for Lv2 and then see what I need to change. Wish me luck :smile: 

@michaelgregorius As for your PR, isn't this two separate things in one PR? Do you think it would make sense to split this PR into "add maximize button" and "fix resizing behavior"?

michaelgregorius commented:
> As my PR from #7524 does not fully work after the SlicerT update and conflict with your PR, I will try to use your PR for Lv2 and then see what I need to change. Wish me luck 😄

Good luck then! :sweat_smile:
> 
> @michaelgregorius As for your PR, isn't this two separate things in one PR? Do you think it would make sense to split this PR into "add maximize button" and "fix resizing behavior"?

Yes, it can definitively be separated as you have described. Therefore I have now separated the sub window fixes into pull request https://github.com/LMMS/lmms/pull/7530. Do you want to review it?

JohannesLorenz commented:
> Do you want to review it?

Yes, thanks, I added myself to the reviewer list.

I'm already in the middle of testing, most things look good.

JohannesLorenz commented on src/gui/instrument/InstrumentTrackWindow.cpp:
```diff
@@ -285,24 +285,30 @@ InstrumentTrackWindow::InstrumentTrackWindow( InstrumentTrackView * _itv ) :
+		subWin->setMaximumSize(m_instrumentView->maximumSize() + extraSpace);
+		subWin->setMinimumSize(m_instrumentView->minimumSize() + extraSpace);
+
+		flags |= Qt::WindowMaximizeButtonHint;
```
I assume, for symmetry reasons to the `else` case, you could clear the `Qt::MSWindowsFixedSizeDialogHint` here, too?

michaelgregorius replied:
Done with commit e85699f7872.

JohannesLorenz reviewed (commented):
What I did:
* First functional review
* Testing review

What needs to be done:
* Style review
* Second functional review (IMO optional)

michaelgregorius commented:
> Issue during testing:
> 
>     * New project
> 
>     * Create SlicerT (don't maximize)
> 
>     * Drag TripleOsc right into the SlicerT window
> 
> 
> Problems:
> 
>     * TripleOsc has wrong size => Not fault of this PR
> 
>     * TripleOsc has maximize button => Should IMO be fixed in this PR
> 
> 
> Variations:
> 
>     * Swap around TripleOsc and SlicerT
> 
>     * Ignore the "(don't maximize)"
> 
> 
> => In all cases, the maximize button should be updated to the new instrument => A maximized window should, after the instrument changes, only be maximized if the new instrument supports it

@JohannesLorenz, all issues except for the "TripleOsc has wrong size => Not fault of this PR" should be fixed by the last three commits.

JohannesLorenz commented on src/gui/instrument/InstrumentTrackWindow.cpp:
```diff
@@ -717,5 +698,66 @@ void InstrumentTrackWindow::adjustTabSize(QWidget *w)
+	return nullptr;
+}
+
+void InstrumentTrackWindow::updateSubWindowState()
```
What I find strange about this method: It is no way specific to Instrument Track SubWindows (except the "isResizable" which I plan to generalize to both Instruments and Effects, i.e. to `PluginView`, but not all `SubWindow` types - nonetheless, `SubWindow` might just do a `qobject_cast` on its children, and if it's a PluginView, use the `isResizable` from there).

I guess your PR is only required for instruments, because they are the only SubWindow type that can be maximized and can both be resizable or not? In this case, the issue will occur for Effects sooner or later. And for the other SubWindow types, I assume it might not harm?

The advantage of putting that code into `SubWindow` would be cleaner encapsulation and no need to find the parent.

michaelgregorius replied:
Resolving this conversation as it was moved out into the main thread with this comment: https://github.com/LMMS/lmms/pull/7514#issuecomment-2406838814

JohannesLorenz commented on src/gui/instrument/InstrumentTrackWindow.cpp:
```diff
@@ -710,5 +698,75 @@ void InstrumentTrackWindow::adjustTabSize(QWidget *w)
+
+		subWindow->setWindowFlags(flags);
+
+		// Show or gide the Size and Maximize options from the system menu depending on whether the view is resizable or not
```
```suggestion
		// Show or hide the Size and Maximize options from the system menu depending on whether the view is resizable or not
```

michaelgregorius replied:
Fixed with commit 8b4976eaaf1.

JohannesLorenz reviewed (commented):
What I did:

* First functional review

What needs to be done:

* Testing review (I volunteer for it once the code is final)
* Style review

michaelgregorius commented:
> What I find strange about this method: It is no way specific to Instrument Track SubWindows (except the "isResizable" which I plan to generalize to both Instruments and Effects, i.e. to `PluginView`, but not all `SubWindow` types - nonetheless, `SubWindow` might just do a `qobject_cast` on its children, and if it's a PluginView, use the `isResizable` from there).
> 
> I guess your PR is only required for instruments, because they are the only SubWindow type that can be maximized and can both be resizable or not? In this case, the issue will occur for Effects sooner or later. And for the other SubWindow types, I assume it might not harm?
> 
> The advantage of putting that code into `SubWindow` would be cleaner encapsulation and no need to find the parent.

Moving this discussion out of the code thread, @JohannesLorenz.

Yes, it should definitively be generalized.

However, with the current hierarchy this would not really work well (I think you already mentioned some of the problems somewhere else). With regards to the instruments the `PluginView` only describes the instrument specific widget, e.g. the orange 3OSC window, via `InstrumentView` which inherits from `PluginView`. However, the child of the `SubWindow` is the `InstrumentTrackWindow` which described the full window with all tabs, etc. and which has rather unspecific inheritances (`QWidget`, `ModelView` and `SerializingObjectHook`).

So it seems that some more specific class would have to be inserted that `InstrumentTrackWindow` and whatever class deals with the effects can inherit from. Let's call it `PluginMainView` for now. It might even be that the `isResizable` method would then need to be moved into `PluginMainView` so that the sub window can query it (assuming that `PluginMainView` is the direct child of the sub window).

Regarding moving logic into the sub window. In that case I'd prefer a sub window class like `PluginSubWindow` which would inherit from `SubWindow` and receive a `PluginMainView` into its constructor and which would then operate on this instance. That way `SubWindow` would only deal with the technical aspects, e.g. how to render itself, whereas `PluginSubWindow` would be used for functional or "business" aspects.

I am still not sure why `SubWindow` exists though? What problems is it expected to solve? I am asking because due to its imperfect implementation it is also causing lots of problems. See for example the public method `updateTitleBar` which had to be introduced because it does not update automatically whereas a `QMdiSubWindow` does.

michaelgregorius commented on src/gui/instrument/InstrumentTrackWindow.cpp:
```diff
@@ -710,5 +698,75 @@ void InstrumentTrackWindow::adjustTabSize(QWidget *w)
+	return nullptr;
+}
+
+void InstrumentTrackWindow::updateSubWindowState()
```
I guess it depends on what's the expected scope of that method. Nevertheless, I have renamed the method with commit a594002d982.

michaelgregorius commented on src/gui/instrument/InstrumentTrackWindow.cpp:
```diff
@@ -710,5 +698,75 @@ void InstrumentTrackWindow::adjustTabSize(QWidget *w)
+
+			// The sub window might be reused from an instrument that was maximized. Show the sub window
+			// as normal, i.e. not maximized, if the instrument view is not resizable.
+			if (subWindow->isMaximized())
```
I guess it's a taste thing but I find the code to be more readable when it's not inlined even if the scope just contains one line.

JohannesLorenz commented:
> I am still not sure why SubWindow exists though? What problems is it expected to solve?

Only to that specific question: Did you see the comment in the header (above the class) and especially in the cpp above `SubWindow::moveEvent`?

Wallacoloo commented:
That's 9 years ago, patching over a bug that existed in QMdiSubWindow 
at the time. If you're able to remove SubWindow and everything still 
appears to work, then it's probably been fixed upstream and is fine to 
remove.

As far as the behavior to check for: does any combination of moving, 
resizing, and maximizing/minimizing a QMdiSubWindow (or the main LMMS 
window) cause the QMdiSubWindow to take on unexpected dimensions? For 
example, maximize the subwindow, then resize the main LMMS window (just 
a pixel or two), then un-maximize the subwindow: does it restore to its 
original size? There was some interaction (probably not exactly that, 
but like that) where it would lose track of its non-maximized 
dimensions and not restore to the size/location you'd expect it to.

On Fri, Oct 11 2024 at 11:03:19 AM -07:00:00, Johannes Lorenz 
***@***.***> wrote:
>> I am still not sure why SubWindow exists though? What problems is it 
>> expected to solve?
>> 
> Only to that specific question: Did you see the comment in the header 
> (above the class) and especially in the cpp above 
> SubWindow::moveEvent?
> 
> See commit 7cc917c 
> <https://github.com/LMMS/lmms/commit/7cc917cc06e44b09378d4e4e166f2dbf148a1e27> 
> . The author @Wallacoloo <https://github.com/Wallacoloo> might want 
> to say something about it, too.
> 
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/LMMS/lmms/pull/7514#issuecomment-2407890586>, or 
> unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAJHS7YLZ4UO7BG7HNUHK7TZ3AHGPAVCNFSM6AAAAABOSZJRZSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMBXHA4TANJYGY>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>

Wallacoloo commented:
Oh, well look at the bug linked at the top of that file (thank you 
whoever linked it): <https://bugreports.qt.io/browse/QTBUG-256>

It's closed, as "out of scope". No activity or Qt commits linking to 
it. It's probably still something you have to deal with on Qt+X11 
(doesn't mention anything about Wayland).

On Fri, Oct 11 2024 at 08:18:06 PM +00:00:00, ***@***.*** wrote:
> That's 9 years ago, patching over a bug that existed in QMdiSubWindow 
> at the time. If you're able to remove SubWindow and everything still 
> appears to work, then it's probably been fixed upstream and is fine 
> to remove.
> 
> As far as the behavior to check for: does any combination of moving, 
> resizing, and maximizing/minimizing a QMdiSubWindow (or the main LMMS 
> window) cause the QMdiSubWindow to take on unexpected dimensions? For 
> example, maximize the subwindow, then resize the main LMMS window 
> (just a pixel or two), then un-maximize the subwindow: does it 
> restore to its original size? There was some interaction (probably 
> not exactly that, but like that) where it would lose track of its 
> non-maximized dimensions and not restore to the size/location you'd 
> expect it to.
> 
> On Fri, Oct 11 2024 at 11:03:19 AM -07:00:00, Johannes Lorenz 
> ***@***.***> wrote:
>>> I am still not sure why SubWindow exists though? What problems is 
>>> it expected to solve?
>>> 
>> Only to that specific question: Did you see the comment in the 
>> header (above the class) and especially in the cpp above 
>> SubWindow::moveEvent?
>> 
>> See commit 7cc917c 
>> <https://github.com/LMMS/lmms/commit/7cc917cc06e44b09378d4e4e166f2dbf148a1e27> 
>> . The author @Wallacoloo <https://github.com/Wallacoloo> might want 
>> to say something about it, too.
>> 
>> —
>> Reply to this email directly, view it on GitHub 
>> <https://github.com/LMMS/lmms/pull/7514#issuecomment-2407890586>, or 
>> unsubscribe 
>> <https://github.com/notifications/unsubscribe-auth/AAJHS7YLZ4UO7BG7HNUHK7TZ3AHGPAVCNFSM6AAAAABOSZJRZSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMBXHA4TANJYGY>.
>> You are receiving this because you were mentioned.Message ID: 
>> ***@***.***>
>>

michaelgregorius commented:
Thanks for the heads up, @Wallacoloo! I could not reproduce any of the problems while changing the state of the sub windows and the main window (under X11). However, I guess switching to `QMdiSubWindow` now would likely open another can of worms anyway as it does not seem to support style sheets very well.

JohannesLorenz commented:
Regarding the other proposals: This is all a bit complex, and might change drastically as soon as I change the hierarchy of `PluginView`/`EffectControlDialog`/`EffectView`. So I suggest either waiting until I have a PR here, or finish this PR as it is (open comment to solve, outstand test review). Ultimately, the decision is up to the author - I would accept both.

michaelgregorius commented:
> Regarding the rest of the Subwindow generalization discussion: This is all a bit complex, and might change drastically as soon as I change the hierarchy of `PluginView`/`EffectControlDialog`/`EffectView`. So I suggest either waiting until I have a PR here, or finish this PR as it is (open comment to solve, outstand test review). Ultimately, the decision is up to the author - I would accept both.

I'd say lets first do the hierarchy changes, hoping that they might make it easier or clearer how to solve the remaining problems with this PR.

JohannesLorenz commented:
> lets first do the hierarchy changes

I uploaded them into a PR now: #7544 .

SpomJ commented on src/gui/instrument/InstrumentTrackWindow.cpp:
```diff
@@ -710,5 +698,75 @@ void InstrumentTrackWindow::adjustTabSize(QWidget *w)
+			subWindow->setMaximumSize(m_instrumentView->maximumSize() + extraSpace);
+			subWindow->setMinimumSize(m_instrumentView->minimumSize() + extraSpace);
+
+			flags &= ~Qt::MSWindowsFixedSizeDialogHint;
```
Consider also adding `setFixedSize()` here. `MSWindowsFixedSizeDialogHint` annoyingly appears to be a windows-only thing, or at least, isn't respected on wayland, while `setFixedSize()` or `layout->setSizeConstraint(QLayout::SetFixedSize` is.

michaelgregorius replied:
@SpomJ , do you mean to set the `subWindow` to a fixed size? If yes, then I am not sure if that makes sense. The calls to `setMaximumSize` and `setMinimumSize` make the sub window resizable according to the limits as defined by the instrument view. Setting a fixed size would take that away. And there would also be the question which fixed size to set?

JohannesLorenz replied:
I agree that this is counter productive. But anyways - this PR is about maximizing, not about min/max size (though the diff might look like it at first sight) - min/max size is a job which #7524.

michaelgregorius replied:
Resolving this conversation then. @SpomJ, feel free to reopen it if necessary.

JohannesLorenz commented:
IMO the comment from @SpomJ can be ignored, as this would, if even, be handled in #7524 (see my comment to their comment). We are also a bit under time pressure (sorry, partially my fault), because this PR could influence #7201 .

> > Regarding the rest of the Subwindow generalization discussion: This is all a bit complex, and might change drastically as soon as I change the hierarchy of `PluginView`/`EffectControlDialog`/`EffectView`. So I suggest either waiting until I have a PR here, or finish this PR as it is (open comment to solve, outstand test review). Ultimately, the decision is up to the author - I would accept both.
> 
> I'd say lets first do the hierarchy changes, hoping that they might make it easier or clearer how to solve the remaining problems with this PR.

@michaelgregorius We were stopping here with the review: https://github.com/LMMS/lmms/pull/7514#issuecomment-2408861562 . Since my hiearchy changes are not going to be realized we should continue there: There are still open comments from the previous review, and then the PR must be tested. Does that make sense?

michaelgregorius commented on src/gui/instrument/InstrumentTrackWindow.cpp:
```diff
@@ -710,5 +698,75 @@ void InstrumentTrackWindow::adjustTabSize(QWidget *w)
+	while (p != nullptr)
+	{
+		auto mdiSubWindow = dynamic_cast<QMdiSubWindow*>(p);
+		if (mdiSubWindow)
```
As this is only a style issue which is a matter of taste I'd rather keep it as is. I find it more readable and due to the braces it is also not very prone to "Heartbleed-like" issues if code is added.

michaelgregorius commented:
@JohannesLorenz, I have merged the current master and did a quick test. Maximization is still working. Is there anything else to do?

michaelgregorius commented:
As noted before this PR fixes some general problems with maximizing Instrument windows correctly. IMO we can merge it and if the maximization feature is not wanted then it would suffice to only remove the code that does something with the `Qt::WindowMaximizeButtonHint` flag so that the feature can be added back quickly and in a working state if it is wanted.

JohannesLorenz reviewed (approved):
I think this can be merged

JohannesLorenz commented:
@michaelgregorius Sorry, abusing this PR for contacting you. I am just going through the conditions to resizability of `InstrumentView`s (mainly for #7524), and I wonder if our idea to introduce a virtual `isResizable(...)` is good. Shouldn't we just use the existing way of letting all `InstrumentView` child classes calling `setSizePolicy(...)` instead of constructing a bypassing, redundant system to transfer this information to the parents?

Btw, I remember you once wrote how I can contact you? Sorry, I do not find the info anymore. If you want to discuss, maybe an instant messenger might be better than reviving an old PR :laughing: Though I don't have a strong opinion here, either.

michaelgregorius commented:
@JohannesLorenz, if this does not break anything else then it might also be worth a try. I guess this would mean that the method `isResizable` would be removed from `EffectControlDialog`, `PluginView` and all inheriting classes. Both base classes would then call something like `setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed)` in their constructors.

Every class that now overrides `isResizable` by returning `true` would then need to call `setSizePolicy` with some other policies than `Fixed`. And the test if a widget is resizable would turn into checking the respective size policy.

The code in `EffectView` might perhaps be changed to simply pass on the size policy of the child to the sub window.

Only thing I am not sure about is in how far the size policy of a widget is specific to layout contexts or if it is a general property. However, I guess it is also evaluated if a widget is for example a top-level widget that's not inside a layout. So it might be general anyway.

I think I proposed to discuss things in Jitsi Meet because I assumed a "real" meeting. However, I cannot find that issue/comment anymore, either. :sweat_smile: Alternatively you can always ping me here in the relevant issues.

JohannesLorenz commented:
@michaelgregorius I have just read a lof of Qt stuff. Finally, (I think) I found out: While `QWdiget::setFixedSize` decides whether the widget can resize after creation, `QWidget::setSizePolicy` decides how the widget behaves in case of a resize, which can happen both at widget initialization or user resize.

> However, I guess it is also evaluated if a widget is for example a top-level widget that's not inside a layout. So it might be general anyway.

When I tried to set the sizePolicy of a `QMdiSubWindow` to `Fixed`, it was still resizable. From what I read, I would even conclude that `QWidget::setSizePolicy` is irrelevant for a `QMdiSubWindow`, because it is a toplevel widget...

> Every class that now overrides isResizable by returning true would then need to call setSizePolicy with some other policies than Fixed. And the test if a widget is resizable would turn into checking the respective size policy.

All in all, I guess I will rather use `QWdiget::setFixedSize` in #7524.

> The code in EffectView might perhaps be changed to simply pass on the size policy of the child to the sub window.

I just found out that this seems to work great in `InstrumentTrackWindow` (more or less the analog case).

michaelgregorius commented:
@JohannesLorenz, Qt's use of sizes and sizing is definitively a complex topic so it's easy to get lost in there for hours. :sweat_smile:

I think `setFixedSize` means that a widget gets a fixed size and is always rendered at that size regardless of its content, be it what it paints itself or with regards to its child widgets with their own sizes and policies. A fixed size widget will always have the set size regardless of whether it's a top level widget or a child in a layout.

The size policy on the other hand is mainly used by other "clients" to inquire a widget about its capabilities with regards to making use of available space and its need for space. Example for clients are layout algorithms which need to check how to distribute available space to widgets. So it is mainly something that's interesting for "others".

To me it always feels a bit like one thing is "driven" from the top whereas the other is driven from the bottom.
